doc/tutorial/intro-scripting.rst

```diff
@@ -0,0 +1,670 @@
+for the ``connection`` record, in order to carry around its data. You
+can see which fields an analyzer adds to the ``connection`` object in
+the “redefinitions” section in the script’s documentation - here for
+HTTP. You can see from that section that the HTTP analyzer adds a
+variable ``http_state`` with type ``HTTP::State`` to the ``connection``
+record - thus, we can use it!
+
+Before we use it, since ``c$http_state`` is an optional field, it could
+be necessary to ensure that the ``c$http_state`` field exists before
+using it. If you use an optional field without it being present, that
+would be an error:
+
+   .. code:: console
+
+      expression error in ./test.zeek, line 7: field value missing (c$http_state)
+
+Therefore, we should wrap anything that uses ``http_state`` with a field
+value existence check with ``?$``:
+
+   .. code:: zeek
+
+      event http_entity_data(c: connection, is_orig: bool, length: count, data: string) {
+          if ( c?$http_state ) {
+              c$http_state$entity += data;
+              print c$http$entity;
+          }
+      }
+
+This should get exactly the same results as before. If you want to test
+it, you can use ``delete c$http_state;`` before the ``if`` statement in
+order to make sure it’s not set when it gets to that point, even though
+it always should be.
+
+This prints the information as it is getting collected. Instead, it
+should only print once at the end. For this, we can use the
+``http_end_entity`` event. Remove the `print` that is in
+``http_entity_data`` and move it to the ``http_end_entity`` event:
+
+   .. code:: zeek
+
+      event http_end_entity(c: connection, is_orig: bool, length: count, data: string) {
+          if ( c?$http_state ) {
+              print c$http_state$entity;
+          }
+      }
+
+TODO: Should we also reset $entity here?
+
+Now, it will only print once - at the end of an entity.
+
+TODO: Can entities be nested? I think not but entities.zeek deals with a
+depth and I really don’t want to. :)
+
+There is one more caveat. This gives theoretically unbounded state
+growth, as `entity` has no upper bound. We should introduce an upper
+bound that users can configure. This is easy with redefineable options!
+
+First, we declare the option at the top of the file in an ``export``
+block:
+
+   .. code:: zeek
+
+      export {
+        option max_reassembled_entity_size = 10000 &redef;
+      }
+
+Then, we want to reach exactly that entity size, but never exceed it.
+You can use ``|...|`` around a string to get its size, like
+``|c$http_state$entity|`` will get the length of the string in that
+field. You can do the same to get the size of most containers, like a
+vector. If we subtract it from ``max_reassembled_entity_size``, that
+should be the remaining length:
+
+   .. code:: zeek
+
+      local remaining_available = max_reassembled_entity_size - |c$http_state$entity|;
+      if (remaining_available <= 0) return;
```

Comment:
Drive-by observation: With the given default `max_reassembled_entity_size` will be deduced as a `count`, and `|..|` also returns a `count`, so the subtraction underflows and we can never see a negative value.

A fix would be either to type the option

```zeek
option max_reassembled_entity_size: int = 10000 &redef;
```

or to cast one of the operands to `int` before the subtraction, e.g.,

```zeek
local len: int = |c$http_state$entity|;
local remaining_available = max_reassembled_entity_size - len;
if ( remaining_available <= 0 )
  return;
```

Unfortunately both of these are probably beyond the scope of what you wanted to talk about here.